File: training/action_vocab.py
"""Shared action vocabulary for learned policies."""

from __future__ import annotations

BOARD_WIDTH = 6

ACTION_KEYS = []

for column in range(BOARD_WIDTH):
    ACTION_KEYS.append(f"UP:{column}")
    ACTION_KEYS.append(f"DOWN:{column}")

for column in range(BOARD_WIDTH - 1):
    ACTION_KEYS.append(f"RIGHT:{column}")

for column in range(1, BOARD_WIDTH):
    ACTION_KEYS.append(f"LEFT:{column}")

ACTION_TO_INDEX = {action_key: index for index, action_key in enumerate(ACTION_KEYS)}


def action_index(action_key: str) -> int:
    """Return the fixed action index used by the learned policy."""

    return ACTION_TO_INDEX[action_key]

File: src/ai/action-vocab.js
import { BOARD_WIDTH, ORIENTATIONS } from "../core/constants.js";

export const ACTION_KEYS = [];

for (let column = 0; column < BOARD_WIDTH; column += 1) {
  ACTION_KEYS.push(`${ORIENTATIONS.UP}:${column}`);
  ACTION_KEYS.push(`${ORIENTATIONS.DOWN}:${column}`);
}

for (let column = 0; column < BOARD_WIDTH - 1; column += 1) {
  ACTION_KEYS.push(`${ORIENTATIONS.RIGHT}:${column}`);
}

for (let column = 1; column < BOARD_WIDTH; column += 1) {
  ACTION_KEYS.push(`${ORIENTATIONS.LEFT}:${column}`);
}

export const ACTION_INDEX = Object.freeze(
  Object.fromEntries(ACTION_KEYS.map((actionKey, index) => [actionKey, index])),
);
